agents/systems/WIDGET-INTEGRATION-GUIDE.md

```diff
@@ -0,0 +1,678 @@
+# Widget Integration Developer Guide
+
+This guide explains how to create widgets that integrate with third-party services using the Castmill Widget Integration System.
+
+## Table of Contents
+
+1. [Overview](#overview)
+2. [Integration Types](#integration-types)
+3. [Credential Scopes](#credential-scopes)
+4. [Creating an Integration](#creating-an-integration)
+5. [PULL Mode Integration](#pull-mode-integration)
+6. [PUSH Mode Integration](#push-mode-integration)
+7. [Player Implementation](#player-implementation)
+8. [Testing](#testing)
+9. [Examples](#examples)
+
+## Overview
+
+The Widget Integration System provides a standardized way for widgets to connect with external services. It handles:
+
+- **Credential Management**: Secure storage and encryption of API keys and credentials
+- **Data Fetching**: Automated PULL of data from third-party APIs
+- **Webhook Handling**: PUSH of data from third-party services
+- **Version Control**: Efficient polling with version-based caching
+- **Error Handling**: Graceful degradation when services are unavailable
+
+## Integration Types
+
+### PULL Mode
+
+The backend periodically fetches data from a third-party API and caches it. Players poll for updates.
+
+**Use Cases:**
+- Weather data
+- Stock prices
+- RSS feeds
+- Social media feeds with API access
+
+**Pros:**
+- Centralized rate limiting
+- Efficient for multiple players
+- Can cache and serve stale data
+
+**Cons:**
+- Delayed updates (limited by pull interval)
+- Backend resource usage
+
+### PUSH Mode
+
+Third-party services push data to a webhook when it changes. Players poll for the latest version.
+
+**Use Cases:**
+- Real-time event notifications
+- Social media webhooks
+- Form submissions
+- IoT sensor data
+
+**Pros:**
+- Near real-time updates
+- No polling of third-party APIs
+- Event-driven architecture
+
+**Cons:**
+- Requires third-party to support webhooks
+- Need to configure webhook URLs externally
+
+### BOTH Mode
+
+Supports both PULL and PUSH. Use PULL for initial data and periodic fallback, PUSH for real-time updates.
```

Comment:
Not sure I can imagine a case where both modes will be used at the same time, I think it is better to only allow one or the other for a given widget.